### PR TITLE
Add support for non-compatible versions

### PR DESCRIPTION
**Pull request by DavidSchinazi**

### CODE REVIEW

File: draft-schinazi-quic-version-negotiation.md [hunk 1]
```diff
@@ -28,78 +28,81 @@ author:
...
-validate that the version in the server's packets is one of the versions that it
-offered and that it matches the value in the server's transport parameters.
+is sent using the version that the server is most likely to support (in the
+absence of other information, this will often be the oldest version the client
+supports); that Initial packet then lists all of the other compatible versions
+({{compatible-versions}}) that the client supports in the Compatible Version
+fields of its transport parameters ({{fig-client-tp}}).
+
+1. If the server supports the client's attempted version, and supports one of
+   the client's compatible versions, it selects the first version it supports
```
Comment (ekr) [inline draft-schinazi-quic-version-negotiation.md]: Why the "first version" restriction rather than one of them? What happens if the client lists them in goofy order?

File: draft-schinazi-quic-version-negotiation.md [hunk 2]
```diff
@@ -28,78 +28,81 @@ author:
...
+   from the client's compatible version list. It then responds responds with
```
Comment (ekr) [inline draft-schinazi-quic-version-negotiation.md]: "respond responds"

File: draft-schinazi-quic-version-negotiation.md [hunk 3]
```diff
@@ -28,78 +28,81 @@ author:
...
-responds with that version in all of its future packets (except for Retry, as
-below). It also inserts the selected version in the
-negotiated_compatible_version field of its transport parameters.
-
-The server MUST NOT select a version not offered by the client.  The client MUST
-validate that the version in the server's packets is one of the versions that it
-offered and that it matches the value in the server's transport parameters.
+is sent using the version that the server is most likely to support (in the
+absence of other information, this will often be the oldest version the client
+supports); that Initial packet then lists all of the other compatible versions
```
Comment (ekr) [inline draft-schinazi-quic-version-negotiation.md]: I know I had "all the other" here, but I think that was stupid. It should just say "all"

File: draft-schinazi-quic-version-negotiation.md [hunk 4]
```diff
@@ -28,78 +28,81 @@ author:
...
+fields of its transport parameters ({{fig-client-tp}}).
+
+1. If the server supports the client's attempted version, and supports one of
+   the client's compatible versions, it selects the first version it supports
+   from the client's compatible version list. It then responds responds with
+   that version in all of its future packets (except for Retry, as below).
+
+1. If the server supports the client's attempted version but does not support
+   any of the client's compatible versions, it proceeds with the clients
+   attempted version.
```
Comment (ekr) [inline draft-schinazi-quic-version-negotiation.md]: See above.

File: draft-schinazi-quic-version-negotiation.md [hunk 5]
```diff
@@ -28,78 +28,81 @@ author:
...
+
+1. If the server does not support the client's attempted version, it sends a
+   Version Negotiation packet listing all the versions it supports.
```
Comment (ekr) [inline draft-schinazi-quic-version-negotiation.md]: I would move this up in the doc because it's going to happen before you read the other stuff.

File: draft-schinazi-quic-version-negotiation.md [hunk 6]
```diff
@@ -28,78 +28,81 @@ author:
...
+
+If the server leverages compatible versions and responds with a different
+version from the client's attempted version, it MUST NOT select a version not
+offered by the client.  The client MUST validate that the version in the
+server's packets is either its attempted version, or one of the compatible
+versions that it offered and that it matches the negotaited version in the
```
Comment (ekr) [inline draft-schinazi-quic-version-negotiation.md]: misspell of negotiated. Also, see above about having the client's list being exhaustive.

File: draft-schinazi-quic-version-negotiation.md [hunk 7]
```diff
@@ -112,42 +115,143 @@ versions as long as they are reasonably popular.  This is not generally an issue
...
+~~~
+{: #fig-client-tp title="Client Transport Parameter Format"}
+
+The content of each field is described below:
+
+Attempted Version:
+
+: The version that the client is using in this Initial. This field MUST be
+  equal to the value of the Version field in the long header that carries
+  this transport parameter.
```
Comment (ekr) [inline draft-schinazi-quic-version-negotiation.md:157]: Why is this needed?

File: draft-schinazi-quic-version-negotiation.md [hunk 8]
```diff
@@ -112,42 +115,143 @@ versions as long as they are reasonably popular.  This is not generally an issue
...
+: A variable-length integer specifying the number of Supported Version fields
+  following it. The server encodes all versions it supports in the subsequent
+  list, ordered by descending preference.
+
+
+Clients MAY include versions following the pattern `0x?a?a?a?a` in their
+Compatible Version list, and the server in their Supported Version list.
+Those versions are reserved to exercise version negotiation (see the Versions
+section of {{QUIC}}), and MUST be ignored when parsing these fields. Clients
+MUST NOT add reserved versions that it did not oberve in the Version
```
Comment (ekr) [inline draft-schinazi-quic-version-negotiation.md]: observe.

File: draft-schinazi-quic-version-negotiation.md [hunk 9]
```diff
@@ -112,42 +115,143 @@ versions as long as they are reasonably popular.  This is not generally an issue
...
+Negotiation packet to the Received Negotiation Version list, and clients MUST
+include any reserved versions seen in the Version Negotiation packet to the
+Received Negotiation Version list.
```
Comment (ekr) [inline draft-schinazi-quic-version-negotiation.md]: Would it be simpler to just say that it must be identical?

File: draft-schinazi-quic-version-negotiation.md [hunk 10]
```diff
@@ -112,42 +115,143 @@ versions as long as they are reasonably popular.  This is not generally an issue
...
+
+# Version Downgrade Prevention
+
+When a server parses the client's `version_negotiation` transport parameter, if
+the `Received Negotiation Version Count` is not zero, the server MUST validate
+that it could have sent the Version Negotation packet described by the client.
+In particular, the server MUST ensure that there are no versions that it
+supports that are absent from the Received Negotiation Versions. This mitigates
+an attacker's ability to forge Version Negotiation packets to force a version
+downgrade.
```
Comment (ekr) [inline draft-schinazi-quic-version-negotiation.md]: Is anything else required.

File: draft-schinazi-quic-version-negotiation.md [hunk 11]
```diff
@@ -112,42 +115,143 @@ versions as long as they are reasonably popular.  This is not generally an issue
...
+
+If a server operator is progressively deploying a new QUIC version throughout
+its fleet, it MAY perform a two-step process where it first progressively adds
+support for the new version, but without enforcing its presence in Received
+Negotiation Versions. Once all servers have been upgraded, the second step is
+to start enforcing that the new version is present in Received Negotiation
+Versions. This opens connections to downgrade attackss during the upgrade
```
Comment (ekr) [inline draft-schinazi-quic-version-negotiation.md]: attacks

**Review by ekr — CHANGES_REQUESTED** [review]
This looks generally like what we agreed. I believe that there are two things that should be changed:

1. I would require the client's compatible versions to all be present in compatible_versions. It's just simpler logic for the server.
2. You need to require the client to check to make sure that it didn't use a version in the server's VN, otherwise attacker may be able to influence you and the server won't detect it.

File: draft-schinazi-quic-version-negotiation.md (hunk 1 above)
Comment (DavidSchinazi) [reply draft-schinazi-quic-version-negotiation.md]: Fixed

File: draft-schinazi-quic-version-negotiation.md (hunk 2 above)
Comment (DavidSchinazi) [reply draft-schinazi-quic-version-negotiation.md]: Fixed

File: draft-schinazi-quic-version-negotiation.md (hunk 3 above)
Comment (DavidSchinazi) [reply draft-schinazi-quic-version-negotiation.md]: Fixed

File: draft-schinazi-quic-version-negotiation.md (hunk 6 above)
Comment (DavidSchinazi) [reply draft-schinazi-quic-version-negotiation.md]: Fixed

File: draft-schinazi-quic-version-negotiation.md (hunk 7 above)
Comment (DavidSchinazi) [reply draft-schinazi-quic-version-negotiation.md:157]: It prevents a class of attacks where the CRYPTO frame is copied from one version to another, by adding it to the TLS transcript. See Security Considerations.

File: draft-schinazi-quic-version-negotiation.md (hunk 8 above)
Comment (DavidSchinazi) [reply draft-schinazi-quic-version-negotiation.md]: Fixed

File: draft-schinazi-quic-version-negotiation.md (hunk 9 above)
Comment (DavidSchinazi) [reply draft-schinazi-quic-version-negotiation.md]: Fixed

File: draft-schinazi-quic-version-negotiation.md (hunk 10 above)
Comment (DavidSchinazi) [reply draft-schinazi-quic-version-negotiation.md]: For a version downgrade to be successful, the attacker needs to forge one of
- client's first attempted version
- server's VN packet
- client's second attempted version
- server's negotiated version

This puts all of these in the TLS transcript so as long as they are validated we should be fine?

File: draft-schinazi-quic-version-negotiation.md (hunk 11 above)
Comment (DavidSchinazi) [reply draft-schinazi-quic-version-negotiation.md]: Fixed

File: draft-schinazi-quic-version-negotiation.md (hunk 5 above)
Comment (DavidSchinazi) [reply draft-schinazi-quic-version-negotiation.md]: Refactored.

**Comment by DavidSchinazi** [comment]
Addressed comments.

File: draft-schinazi-quic-version-negotiation.md (hunk 7 above)
Comment (ekr) [reply draft-schinazi-quic-version-negotiation.md:157]: Did you forget to write that text? I dont' see it.

File: draft-schinazi-quic-version-negotiation.md [hunk 12]
```diff
@@ -112,42 +111,148 @@ versions as long as they are reasonably popular.  This is not generally an issue
...
+Clients MUST ignore any received Version Negotiation packets that contain the
+version that they initially attempted.
+
+When a server parses the client's `version_negotiation` transport parameter, if
+the `Received Negotiation Version Count` is not zero, the server MUST validate
+that it could have sent the Version Negotation packet described by the client.
+In particular, the server MUST ensure that there are no versions that it
+supports that are absent from the Received Negotiation Versions, and that the
+ordering matches the server's preference. This mitigates an attacker's ability
+to forge Version Negotiation packets to force a version downgrade.
```
Comment (ekr) [inline draft-schinazi-quic-version-negotiation.md]: Suppose the client has preferences 3 > 2 > 1. However because
it thinks 3 is rare, it tends to first try 2. The attacker
supports 3 > 2 > 1.

In the ordinary case, the client would offer 2 and the server
would accept it.

However, consider what happens if the attacker forges a VN:

~~~
C->S: Ver = 2, Attempted Version = 2, VN = []
A->C: VN = 3, 2, 1
C->S: Ver = 3, Attempted Version = 3, VN = [3, 2, 1]
~~~
This handshake will succeed, and the attacker has succeeded
in *upgrading* you. I know that this doesn't seem like a real
attack, but formally they have influenced the negotiation.

File: draft-schinazi-quic-version-negotiation.md (hunk 7 above)
Comment (DavidSchinazi) [reply draft-schinazi-quic-version-negotiation.md:157]: Good point, added normative validation text

File: draft-schinazi-quic-version-negotiation.md (hunk 12 above)
Comment (DavidSchinazi) [reply draft-schinazi-quic-version-negotiation.md]: This wouldn't work because the client would drop a VN packet containing the version it tried. But I've added the previously attempted version field to tighten this even more.
